lib/ide/src/double_representation.rs

```diff
@@ -0,0 +1,79 @@
+//! A module with all functions used to synchronize different representations of our language
+//! module.
+
+use ast::IdMap;
+use data::text::Size;
+use data::text::Span;
+
+
+
+// ================
+// === Text API ===
+// ================
+
+/// Update IdMap to reflect the recent code change.
+pub fn apply_code_change_to_id_map(id_map:&mut IdMap, removed:&Span, inserted:&str) {
+    let IdMap(vector) = id_map;
+    let inserted_len  = Size::new(inserted.chars().count());
+    vector.drain_filter(|(span,_)| removed.contains_span(&span));
+    for (span, _) in vector {
+        if span.index >= removed.end() {
+            span.index += inserted_len;
+            span.index -= removed.size;
+        } else if span.index >= removed.index {
+            let removed_chars = removed.end() - span.index;
+            span.index = removed.index + inserted_len;
+            span.size -= removed_chars;
+        } else if span.end() >= removed.index {
+            let removed_chars = (span.end() - removed.index).min(removed.size);
+            span.size -= removed_chars;
+            span.size += inserted_len;
+        }
+    }
+}
+
+
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use ast::IdMap;
+    use data::text::Index;
+    use data::text::Size;
+    use data::text::Span;
+    use uuid::Uuid;
+
+    #[test]
+    fn applying_code_changes_to_id_map() {
+        let uuid1 = Uuid::new_v4();
+        let uuid2 = Uuid::new_v4();
+        let uuid3 = Uuid::new_v4();
+        let uuid4 = Uuid::new_v4();
+        let uuid5 = Uuid::new_v4();
+        let mut id_map = IdMap(vec!
+            [ (Span::new(Index::new(0) , Size::new(3)), uuid1)
```

Comment:
Not a request for change — it would take more time and perhaps it's not worthy of that at all…  but this indices are to me really unreadable and unmaintainable. I have no idea what's going on without spending much time on counting the characters.

In somewhat similar situation in Scala I've introduced a mini-dsl notation for program sourcecode that allowed annotating it with necessary information.